rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,426 @@
+
+     Type: const :: a -> b -> a
+     Example:
+       let f = const 5; in f 10
+       => 5
+  */
+  const =
+    # Value to return
+    x:
+    # Value to ignore
+    y: x;
+```
+
+## Current problems
+
+### Multiplicity of formats
+
+Within nixpkgs alone, several conventions for doc-comments have emerged, see [1], [2] and [3].
+
+Notably, most doc-comments utilize some fraction of the of CommonMark syntax, even if they are not meant to be rendered.
+
+[1]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/trivial-builders/default.nix
+[2]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/make-derivation.nix
+[3]: https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/make-disk-image.nix
+
+In general, the format for writing documentation strings is **not formally specified**.
+
+Among the formats encountered in the wild, the one used in `nixpkgs/lib` is the only one intended to be rendered as part of an API documentation, via the nixdoc third-party tool, whose syntax has not been standardized.
+
+### Impossible to differentiate from internal comments
+
+The lack of a formal definition of a doc-comment also means there is no reliable way to distinguish them from internal comments, which would result in automatically-produced API documentation which includes the wrong type of comments.
+
+### References to the problems above
+
+This curated link collection highlights the Nix ecosystem's inconsistencies, a primary focus of this RFC.
+
+#### nixpkgs - comment examples
+
+- [lib/attrsets](https://github.com/NixOS/nixpkgs/blob/master/lib/attrsets.nix)
+- [trivial-builders](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/trivial-builders.nix)
+- [stdenv/mkDerivation](https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/make-derivation.nix)
+- [nixos/lib/make-disk-image](https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/make-disk-image.nix)
+
+# Design
+[design]: #detailed-design
+
+In the following we give a comprehensive overview to our decision that we've made. Detailed arguments for and against every decision can be found in the [Decisions](#Decisions) section
+
+## CommonMark
+
+**CommonMark is the content of all doc-comments.**
+
+Adopting CommonMark as the content for all doc-comments brings the benefits of widely accepted and understood documentation format in tech projects, while maintaining profitability and consistency within the Nix ecosystem by aligning with existing [NixOS/RFC-72](https://github.com/NixOS/rfcs/blob/master/rfcs/0072-commonmark-docs.md).
+
+## `/** */` is the doc-comment format
+
+The decision to use /** to start a doc-comment ensures a unique distinction from regular comments while still allowing seamless writing without IDE or editor support. A single choice provides the best developer experience and simplifies tooling.
+
+## Placement
+
+**The placement describes the relationship between doc-comments and the expression they are documenting.**
```

Comment:
I'm not sure if it's the right mental model to think of it as doc comments being attached to expressions. Say you have an attribute set interface where you want to be able to render documentation for individual attributes, but there's a wrapping `if` expression:

```nix
/** (A) */
{ cacheFile }:
if builtins.pathExists cacheFile then
  import cacheFile
else
  {
    /** (B) */
    backend = "bar";
    result = "bar-result";
  }
```

Where would I add a doc-comment to document what `backend` is?
- (A) Putting it here would attach it to the lambda, you'd have to manually write headers or so to indicate attributes, and it couldn't be queried programmatically to belong to a specific attribute.
- (B) Putting it here doesn't feel right. It's not that I'm only documenting the second branch, it's just that there's no literal attribute set in the other branch.

All of this could get really nasty if you have a lot of nestings.

This relates to being able to specify a type for expressions. Notably the module system doesn't suffer from this because there you have a separate DSL to describe the interface and its types.

Reply:
Right i agree totally. In this case i think it would be best to write a doc-comment in place *A*. 
If the function really returns different `attribute sets` depending on the  `ìf` statement there. This means the function has a *non-constant return type*. Which I try to avoid as much as possible because i think it's and anti-pattern. 
But i get the point.

I wouldn't want to use the module system just for documentation. Having documentation as values is an approach which we also considered but i think it is valuable to provide both methods.

Reply:
Ah no the idea here is that both branches return the same type (I tried to illustrate this with one branch just being the cached version). In the first branch it's just read from a different file.

Reply:
@infinisil Are your concerns addressed?